### PR TITLE
Initial implementation of the lua scripting feature

### PR DESCRIPTION
**Pull request by gigaherz**

### CODE REVIEW

**Comment by skmp** [comment]
This pull request **introduces 4 alerts** when merging 0bbb8ad5369292142e46ea2842ea8dd3e8ab5483 into 0afea0edd810a85ffe65fafc1784b9cdb079be80 - [view on LGTM.com](https://lgtm.com/projects/g/reicast/reicast-emulator/rev/pr-0e494dda8131baaa9cf687f7fe60e94fd8132562)

**new alerts:**

* 2 for Sign check of bitwise operation
* 2 for Use of potentially dangerous function

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: libswirl/deps/lua/Makefile [hunk 1]
```diff
@@ -0,0 +1,197 @@
+# Makefile for building Lua
```
Comment (skmp) [inline libswirl/deps/lua/Makefile]: Do we actually use this makefile or is it part of the 3rdparty drop?

File: libswirl/hw/sh4/dyna/driver.cpp [hunk 2]
```diff
@@ -297,7 +297,7 @@ u32 DYNACALL rdv_DoInterrupts_pc(u32 pc) {
 	//We can only safely relocate/etc stuff here, as in other generic update cases
 	//There's a RET, meaning the code can't move around
 	//Interrupts happen at least 50 times/second, so its not a problem ..
-	if (rebuild_counter == 0)
+	//if (rebuild_counter == 0)
```
Comment (skmp) [inline libswirl/hw/sh4/dyna/driver.cpp]: Accidental changes?

File: libswirl/scripting/lua_bindings.cpp [hunk 3]
```diff
@@ -0,0 +1,304 @@
+#include "lua_bindings.h"
+#include "hw/pvr/Renderer_if.h"
+#include "rend/gui.h"
+#include "hw/sh4/sh4_mem.h"
+#include "hw/arm7/arm_mem.h"
+
+extern void dc_loadstate();
```
Comment (skmp) [inline libswirl/scripting/lua_bindings.cpp]: There should be a header file for these

File: libswirl/windows/winmain.cpp [hunk 4]
```diff
@@ -22,6 +22,12 @@
 #include "hw/mem/_vmem.h"
 
 #include "hw/pvr/Renderer_if.h"
+#include <io.h>
```
Comment (skmp) [inline libswirl/windows/winmain.cpp]: Are these actually used?

**Review by skmp — COMMENTED** [review]
Overall looks good for the first PR. See @ comments.

**Comment by skmp** [comment]
There's some code cleanliness that should be done, in the spirit of the refactors for the next part of it.

eg, have a "scripting host" that is always compiled (so that we don't have the #ifdefs), and have some auto-register function register the lua "backend" ?

File: libswirl/deps/lua/Makefile (hunk 1 above)
Comment (gigaherz) [reply libswirl/deps/lua/Makefile]: Part of the 3rdparty package. Had to remove other things, but forgot to remove this file..

File: libswirl/hw/sh4/dyna/driver.cpp (hunk 2 above)
Comment (gigaherz) [reply libswirl/hw/sh4/dyna/driver.cpp]: I think that's a merge derp, but yeah removed.

File: libswirl/windows/winmain.cpp (hunk 4 above)
Comment (gigaherz) [reply libswirl/windows/winmain.cpp]: Leftovers. Removed.

File: libswirl/scripting/lua_bindings.cpp (hunk 3 above)
Comment (gigaherz) [reply libswirl/scripting/lua_bindings.cpp]: Fixed.

**Review by skmp — APPROVED** [review]
Let's merge as soon as it passes CI. Travis might be broken - it's their system that is down atm

**Comment by skmp** [comment]
This pull request **introduces 4 alerts** when merging 07968dd8cb0ef46ecefa8b7c350b2fb4a45cc03c into a52af135946dcf528cf25ce82d809174cc9a99cf - [view on LGTM.com](https://lgtm.com/projects/g/reicast/reicast-emulator/rev/pr-a502d9a0563bb0b1e160d8c85ac1f34d6fcf07d8)

**new alerts:**

* 2 for Sign check of bitwise operation
* 2 for Use of potentially dangerous function

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**